src/HttpServer.php

```diff
@@ -60,50 +65,104 @@ public function addMiddleware($middleware)
-        $bodyCompleted = false;
         $that = $this;
+        $request = null;
+        $headerBuffer = '';
+        $transferredLength = 0;
 
-        $chunkStream = new ReadableStream();
-        $chunkedDecoder = new ChunkedDecoder($chunkStream);
-
-        $headerStream = new ReadableStream();
-        $headerDecoder = new HeaderDecoder($headerStream);
 
-        $connection->on('data', function ($data) use ($connection, &$headerCompleted, &$bodyBuffer, $that, &$chunkedDecoder, &$headerDecoder, $chunkStream, $headerStream) {
-            if (!$headerCompleted) {
-                $headerDecoder->on('data', function ($header) use (&$request, &$data, &$headerCompleted) {
-                    $request = RingCentral\Psr7\parse_request($header);
-                    $data = substr($data, strlen($header));
-                    $headerCompleted = true;
-                });
-                $headerStream->emit('data', array($data));
+        $connection->on('data', function ($data) use (&$connection, $that, &$request, &$headerBuffer, &$transferredLength) {
+            $headerBuffer .= $data;
+            if (strpos($headerBuffer, "\r\n\r\n")) {
+                // header is completed
+                $fullHeader = substr($headerBuffer, 0, strpos($headerBuffer, "\r\n\r\n") + 4);
+                $request = RingCentral\Psr7\parse_request($fullHeader);
+                // remove header from $data, only body is left
+                $data = (string)substr($data, strlen($fullHeader));
+                $input = $that->sendHttpBodyStream($request, $connection);
             }
 
-            if (isset($request)) {
-                if ($that->isChunkedEncodingActive($request)) {
-                    $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $that, $connection) {
-                        $bodyBuffer .= $chunk;
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
-                    }
-                }
+            if ($request != null) {
+                $connection->removeAllListeners('data');
+                $connection->on('data', function ($data) use (&$connection, $that, &$request, &$transferredLength, &$input) {
+                    $that->handleBody($request, $data, $connection, $transferredLength, $input);
```

Comment:
What is input? Why is transferredLength here? This function already receives the connection, so why not process the data events there?